opendcs-rest-api/src/main/java/org/opendcs/odcsapi/res/AppResources.java

```diff
@@ -222,86 +344,166 @@ else if (appStat.getPid() != null && appStat.getPid() != cli.get().getPid())
+		ret.setAppType(app.getAppType());
+		ret.setProperties(app.getProperties());
+		return ret;
+	}
+
 	@POST
 	@Path("appstop")
 	@Consumes(MediaType.APPLICATION_JSON)
 	@Produces(MediaType.APPLICATION_JSON)
 	@RolesAllowed({AuthorizationCheck.ODCS_API_ADMIN, AuthorizationCheck.ODCS_API_USER})
 	public Response postAppStop(@QueryParam("appid") Long appId)
-		throws WebAppException, DbException, SQLException
+			throws WebAppException, DbException
 	{
-		LOGGER.debug("Post appstop received on app with id {}", appId);
 		if (appId == null)
-			throw new WebAppException(ErrorCodes.MISSING_ID, "appId parameter required for this operation.");
-		
-		try (DbInterface dbi = new DbInterface();
-			ApiAppDAO dao = new ApiAppDAO(dbi))
+		{
+			throw new WebAppException(HttpServletResponse.SC_BAD_REQUEST,
+					"appId parameter required for this operation.");
+		}
+
+		try (LoadingAppDAI dai = getLegacyDatabase().makeLoadingAppDAO())
 		{
 			// Retrieve ApiLoadingApp and ApiAppStatus
-			ApiLoadingApp loadingApp = dao.getApp(appId);
-			
-			ApiAppStatus appStat = dao.getAppStatus(appId);
-			
-			if (appStat.getPid() == null)
-				throw new WebAppException(ErrorCodes.NO_SUCH_OBJECT,
-					"appId " + appId + "(" + loadingApp.getAppName() + ") not currently running.");
-			
-			dao.terminateApp(appId);
-			
-			return ApiHttpUtil.createResponse("App with ID " + appId + " (" + loadingApp.getAppName() + ") terminated.");
+			ApiLoadingApp loadingApp = mapLoading(dai.getComputationApp(DbKey.createDbKey(appId)));
+
+			ApiAppStatus appStat = getAppStatus(dai, appId);
+
+			if (appStat == null || appStat.getPid() == null)
+			{
+				throw new WebAppException(HttpServletResponse.SC_NOT_FOUND,
+						"appId " + appId + "(" + loadingApp.getAppName() + ") not currently running.");
+			}
+
+			dai.releaseCompProcLock(new TsdbCompLock(DbKey.createDbKey(appId), appStat.getPid().intValue(),
+					appStat.getHostname(), appStat.getHeartbeat(), appStat.getStatus()));
+
+			return Response.status(HttpServletResponse.SC_OK)
+					.entity("App with ID " + appId + " (" + loadingApp.getAppName() + ") terminated.").build();
+		}
+		catch (NoSuchObjectException e)
+		{
+			return Response.status(HttpServletResponse.SC_NOT_FOUND)
+					.entity("No such app found with ID " + appId).build();
+		}
+		catch (DbIoException ex)
+		{
+			throw new WebAppException(ErrorCodes.DATABASE_ERROR,
+					String.format("Error attempting to stop appId=%s", appId), ex);
+		}
+	}
 
+	private static ApiAppStatus getAppStatus(LoadingAppDAI dai, Long appId) throws DbException
+	{
+		try
+		{
+			List<TsdbCompLock> locks = dai.getAllCompProcLocks();
+			for (TsdbCompLock lock : locks)
+			{
+				if(lock.getAppId().getValue() == appId)
+				{
+					return map(dai, lock);
+				}
+			}
+			List<ApiAppRef> apps = dai.listComputationApps(false)
+					.stream()
+					.map(AppResources::map)
+					.collect(Collectors.toList());
+			for (ApiAppRef app : apps)
+			{
+				if (app.getAppId().equals(appId))
```

Comment:
Can't this be combined with the streaming call on 488?

Reply:
Cleaned up the logic here